datadog/dogstatsd/base.py

```diff
@@ -451,12 +475,21 @@ def _send_to_server(self, packet):
             self.packets_dropped += 1
 
     def _send_to_buffer(self, packet):
-        self.buffer.append(packet)
-        if len(self.buffer) >= self.max_buffer_size:
+        if self._should_flush(len(packet)):
             self._flush_buffer()
+        self.buffer.append(packet)
+        self._current_buffer_total_size += (len(packet) + 1)
+
+    def _should_flush(self, length_to_be_added):
+        if len(self.buffer) >= self._max_buffer_items:
+            return True
+        if self._current_buffer_total_size + length_to_be_added > self._max_payload_size:
+            return True
+        return False
```

Comment:
This will break behavior if the user set a high number for `_max_buffer_items`.

Reply:
Agree, will do as detailed above.